src/message_bus.zig

```diff
@@ -624,57 +629,81 @@ const Connection = struct {
-        );
-    }
-
-    fn recv_body(self: *Connection) void {
-        self.recv(
-            on_recv_body,
-            self.incoming_message.?.buffer[self.recv_progress..self.incoming_message.?.header.size],
-        );
+            return;
+        };
+        self.recv();
     }
 
-    fn recv(
-        self: *Connection,
-        comptime callback: fn (*Connection, *IO.Completion, IO.RecvError!usize) void,
-        buffer: []u8,
-    ) void {
+    fn recv(self: *Connection) void {
         assert(self.peer != .none);
         assert(self.state == .connected);
         assert(self.fd != -1);
+
+        if (self.pipeline_message) |pipeline| {
+            assert(self.recv_progress != 0);
+
+            const buffer = pipeline.buffer[self.pipeline_start..self.recv_progress];
+            if (buffer.len >= @sizeOf(Header)) {
+                const header = mem.bytesAsValue(Header, buffer[0..@sizeOf(Header)]);
+                if (!header.valid_checksum()) {
+                    log.err("invalid checksum on header received from {}", .{self.peer});
+                    self.shutdown();
+                    return;
+                }
+
+                if (buffer.len >= header.size) {
+                    const body = buffer[@sizeOf(Header)..header.size];
+                    if (!header.valid_checksum_body(body)) {
+                        log.err("invalid checksum on body received from {}", .{self.peer});
+                        self.shutdown();
+                        return;
+                    }
+                    mem.copy(u8, self.incoming_message.?.buffer, buffer[0..header.size]);
+
+                    // TODO: rewrite ConcurrentRanges to not use async. This nosuspend is only
+                    // safe because we do not do any asynchronous disk IO yet.
+                    var frame = async self.message_bus.server.on_message(self.incoming_message.?);
+                    nosuspend await frame;
+
+                    self.message_bus.unref(self.incoming_message.?);
+                    self.incoming_message = null;
+
+                    if (header.size == self.recv_progress) {
+                        self.recv_progress = 0;
+                        self.message_bus.unref(pipeline);
+                        self.pipeline_message = null;
+                    }
+                    self.maybe_recv();
+                    return;
+                }
+            }
+            // The pipeline message is not yet fully received, copy to the new buffer and start
+            // a new recv operation.
+            mem.copy(u8, self.incoming_message.?.buffer, buffer);
+            self.recv_progress = buffer.len;
+            self.message_bus.unref(pipeline);
+            self.pipeline_message = null;
+        }
+
         assert(!self.recv_submitted);
         self.recv_submitted = true;
         self.message_bus.io.recv(
             *Connection,
             self,
-            callback,
+            on_recv,
```

Comment:
In the stage1 zig compiler and current language they actually have different types and `self.on_recv` won't compile here as it is a "bound function" not a plain function type. This is however planned to be changed in the language as the migration to the stage2 compiler is made so that the type of `on_recv` and `self.on_recv` are the same. This will make the following no longer work which is IMO a good thing:
```zig
(if (foo) bar else bar2).func();
```
which can of course be rewritten as
```zig
if (foo) bar.func() else bar2.func();
```

Reply:
Thanks for the explanation. That will be a good change to make, and the second example is also cleaner.